### PR TITLE
cylc reload: report added/removed tasks

### LINKED ISSUE #1265: cylc reload: daemon should report added/removed tasks
When a suite is reloaded, for each task in the task pool 'foo' it reports: 
```
RELOADING TASK DEFINITION FOR foo
```
It would be good if it reported additional tasks or removal of tasks as well.

### LINKED ISSUE #1867: Reload after removing a task and its runtime definition fails since cylc-10.0.0
The current reload  remove-task test removes a task from the graph, but not its [runtime] definition.

Test case:
```
[scheduling]
    [[dependencies]]
          graph = foo => bar
[runtime]   
    [[foo]]
    [[bar]]
```

Start suite with `--hold`; remove `bar` from graph and runtime; reload:
```
Processing 1 queued command(s)
  + reload_suite
Traceback (most recent call last):
  File "/home/oliverh/cylc/cylc.git/lib/cylc/scheduler.py", line 236, in start
    self.run()
  File "/home/oliverh/cylc/cylc.git/lib/cylc/scheduler.py", line 1591, in run
    self.update_state_summary()
  File "/home/oliverh/cylc/cylc.git/lib/cylc/scheduler.py", line 1669, in update_state_summary
    self.config.ns_defn_order, self.pool.do_reload)
  File "/home/oliverh/cylc/cylc.git/lib/cylc/network/suite_state.py", line 146, in update
    for parent in ancestors_dict[key]:
KeyError: 'bar'
```

Works fine back at 6.9.1:
```
Processing 1 queued command(s)
  + reload_suite
RELOADING the suite definition
2016-05-26T15:54:00+12 WARNING - orphaned task will not continue: bar.1
```

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #1265 
Closed #1867 

@arjclark Please Review
@hjoliver Please Review

### CODE REVIEW

**Comment by hjoliver** [comment]
Looks good, but while testing I found this: #1867

File: lib/cylc/scheduler.py [hunk 1]
```diff
@@ -1134,6 +1138,21 @@ def load_suiterc(self, reconfigure):
...
+                old_config.get_task_name_list())
+            rem = set(old_config.get_task_name_list()) - set(
+                self.config.get_task_name_list())
+            if len(add) + len(rem) > 0:
+                msgs = ["Added task: '%s'" % task for task in add]
+                msgs += ["Removed task: '%s'" % task for task in rem]
+                for msg in msgs:
+                    logging.getLogger("main").log(logging.INFO, msg)
+                    print msg
+                    print >> sys.stderr, (msg)
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: Don't need the print to stderr?

**Comment by hjoliver** [comment]
Works fine (if the removed-task runtime definition still exists - #1867)

File: lib/cylc/scheduler.py (hunk 1 above)
Comment (arjclark) [inline lib/cylc/scheduler.py]: @hjoliver - it's probably redundant while these messages exist:
```
 2016-05-26T12:26:11+01 WARNING - [bar.1] -last instance (orphaned by reload) 
```
Do we want the message in just suite `log` or `out` file rather than both (probably redundant too) and if so which one?

File: lib/cylc/scheduler.py [hunk 2]
```diff
@@ -1123,9 +1123,8 @@ def configure_pyro(self):
     def load_suiterc(self, reconfigure):
         """Load and log the suite definition."""
 
-        old_config = False
         if reconfigure:
-            old_config = self.config
+            old_tasks = set(self.config.get_task_name_list())
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: This is better, but just out of curiosity - does it actually make a measurable difference to the memory high water mark (or otherwise?)?  I thought about this on my first look, but figured the reference to the full `old_config` would only live for the rest of the function call, and garbage collection isn't going to happen instantly anyway.

File: lib/cylc/scheduler.py (hunk 1 above)
Comment (hjoliver) [inline lib/cylc/scheduler.py]: IMO existing task proxies orphaned by removal of their definitions warrant a logged warning, which gets printed to stderr as well by default. 

New task definitions added probably also warrants a warning, as a reminder that the user has to manually insert new instances if they are to be used.

In which case, no additional printing is needed.

**Comment by hjoliver** [comment]
@oliver-sanders - sorry I didn't notice this before, but the *removed task* part of this PR is somewhat redundant with the existing "orphaned task" handling in `lib/cylc/task_pool.py` - could you consolidate the two?  (And while you're at it, it looks like the `pool.orphans` list only gets house-kept on a reload).

File: lib/cylc/scheduler.py (hunk 2 above)
Comment (arjclark) [inline lib/cylc/scheduler.py]: I requested the removal of this. Primarily it's just better to only grab the bit needed rather than request additional memory allocation for stuff that isn't actually required. In theory you'd get an increased memory footprint here of N + N if you make a copy of the original vs. N + M where N > M for using only the subset of information needed (a list of names rather than their full configs).

I'd also be concerned about further memory spikes during the reload due to keeping the full extra config around. I think mostly there wouldn't be a problem but I pondered whether some of the larger, nastier suites that are starting to appear might be awkward.

TLDR version - probably won't make a noticeable difference but is just a better way to go about it.

**Comment by oliver-sanders** [comment]
@hjoliver @arjclark 
- The 'Removed task' warning is now only logged when an 'orphaned task' warning isn't
- Warnings now only send to the log (level = `WARNING`)

**Comment by hjoliver** [comment]
@oliver-sanders - could you just add `[runtime]` subsections for the added/removed tasks in the new test, so that it tests #1867 too?

**Comment by hjoliver** [comment]
Review 1 - good, over to @arjclark

**Comment by arjclark** [comment]
Looks OK to me. No problems from the test-battery in my environment. Should prove very useful.
